### PR TITLE
Add #[cfg(...)] for select!

### LINKED ISSUE #3974: tokio::select! cfg-gated
**Is your feature request related to a problem? Please describe.**

Select with part of it cfg-gated. Like `#[cfg(not(windows))]`.

**Describe the solution you'd like**

```rust
tokio::select! {
    #[cfg(not(windows))]
    Some(signal) = self.signals.next() => ...
}
```

**Describe alternatives you've considered**

I think the macro could take in a `meta` but that includes other non-cfg stuff, not sure if we want that.

The only workaround is to duplicate the whole block with `if !cfg`.

### PR DESCRIPTION
**Pull request by pickfire**
Fix #3974

## Motivation

Allow conditional compilation in `select!`

## Solution

Parse in `#[cfg($meta:meta)]` in the macro, not sure if we need to do the same for `cfg_attr`.

Currently it fails with recursion limit, not sure if to fix that.

```
error: recursion limit reached while expanding `$crate::select!`
   --> tokio/tests/macros_select.rs:487:15
    |
487 |       let foo = tokio::select! {
    |  _______________^
488 | |         foo = one() => foo,
489 | |         #[cfg(any())]
490 | |         _ = nani() => foo,
491 | |     };
    | |_____^
    |
    = help: consider adding a `#![recursion_limit="256"]` attribute to your crate (`macros_select`)
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

### CODE REVIEW

File: tokio/src/macros/select.rs [hunk 1]
```diff
@@ -562,41 +563,41 @@ macro_rules! select {
     (@ { start=$start:expr; $($t:tt)* } else => $else:expr $(,)?) => {
         $crate::select!(@{ start=$start; $($t)*; $else })
     };
-    (@ { start=$start:expr; ( $($s:tt)* ) $($t:tt)* } $p:pat = $f:expr, if $c:expr => $h:block, $($r:tt)* ) => {
```
Comment (pickfire) [inline tokio/src/macros/select.rs:565]: Should I duplicate this whole block here to fix the recursion limit?

File: tokio/tests/macros_select.rs [hunk 2]
```diff
@@ -547,3 +547,14 @@ pub async fn default_numeric_fallback() {
...
+#[tokio::test]
+async fn select_cfg() {
+    let foo = tokio::select! {
+        foo = one() => foo,
+        #[cfg(any())]
+        _ = nani() => foo,
+    };
+
+    assert_eq!(foo, 1);
+}
```
Comment (Darksonn) [inline tokio/tests/macros_select.rs:560]: I'd like to make sure we are testing both `#[cfg(...)]` queries that enable the branch and `#[cfg(...)]` that disable the branch.

File: tokio/src/macros/select.rs (hunk 1 above)
Comment (Darksonn) [reply tokio/src/macros/select.rs:565]: I don't know how to fix it. You'll have to try it out.

**Comment by pickfire** [comment]
I tried both methods I can think of, one was stuck at `let (_, #[cfg(...)] _, _) = ...` which is not possible. And another one is stuck on attributes on expressions which is not possible yet.

Can I get help for this?

File: tokio/tests/macros_select.rs (hunk 2 above)
Comment (pickfire) [reply tokio/tests/macros_select.rs:560]: Done

**Comment by pickfire** [comment]
Nevermind, I got it working now thanks to guidance from someone on the internet. But implementation wise it isn't zero-cost. Although it will cost some space it will not be polled.

**Comment by pickfire** [comment]
Hmm, the CI seemed to fail on parts I didn't touch.

**Comment by Darksonn** [comment]
It gave this error:
```
---- src/codec/mod.rs - codec (line 169) stdout ----
error: no rules expected the token `}`
  --> src/codec/mod.rs:183:5
   |
16 | /     tokio::select! {
17 | |         num_written = io_resource.write(&buf), if !buf.is_empty() => {
18 | |             buf.advance(num_written?);
19 | |         },
...  |
27 | |         },
28 | |     }
   | |_____^ no rules expected this token in macro call
   |
```
This does indeed seem like something has broken.

**Comment by pickfire** [comment]
Ah, seemed like I missed copying a branch.

File: tokio/tests/macros_select.rs [hunk 3]
```diff
@@ -547,3 +547,24 @@ pub async fn default_numeric_fallback() {
...
+    };
+
+    assert!(foo);
+}
+
+#[tokio::test]
+async fn select_cfg_false() {
+    let foo = tokio::select! {
+        #[cfg(any())]
+        _ = nani() => nani(),
```
Comment (Darksonn) [inline tokio/tests/macros_select.rs:565]:
```suggestion
        _ = this_function_does_not_exist() => this_function_does_not_exist(),
```

File: tokio/src/macros/select.rs [hunk 4]
```diff
@@ -430,7 +449,8 @@ macro_rules! select {
         // This module is defined within a scope and should not leak out of this
         // macro.
         mod util {
-            // Generate an enum with one variant per select branch
+            // Generate an enum with one variant per select branch, including
+            // the branches which cfg is false.
             $crate::select_priv_declare_output_enum!( ( $($count)* ) );
```
Comment (Darksonn) [inline tokio/src/macros/select.rs:454]: I'm pretty sure you can get it to not generate the branch at all if you modify the `select_priv_declare_output_enum` macro to generate output like this:
```Rust
pub(super) enum Out<_0, #[cfg(my_condition)] _1, _2> {
    _0(_0),
    #[cfg(my_condition)]
    _1(_1),
    _2(_2),
    Disabled,
}

pub(super) type Mask = u8;
```
Though you still can't have it "compress" the range of integers to the actual number of enabled ones, so you'll probably need to add an `_ => continue` branch to the match in the loop.

File: tokio/src/macros/select.rs [hunk 5]
```diff
@@ -562,41 +595,65 @@ macro_rules! select {
...
+    (biased; $( #[cfg($meta:meta)] )? $p:pat = $($t:tt)* ) => {
+        $crate::select!(@{ start=0; () } $( #[cfg($meta)] )? $p = $($t)*)
     };
 
-    ( $p:pat = $($t:tt)* ) => {
+    ( $( #[cfg($meta:meta)] )? $p:pat = $($t:tt)* ) => {
         // Randomly generate a starting point. This makes `select!` a bit more
         // fair and avoids always polling the first future.
-        $crate::select!(@{ start={ $crate::macros::support::thread_rng_n(BRANCHES) }; () } $p = $($t)*)
+        $crate::select!(@{ start={ $crate::macros::support::thread_rng_n(BRANCHES) }; () } $( #[cfg($meta)] )? $p = $($t)*)
```
Comment (Darksonn) [inline tokio/src/macros/select.rs:656]: As the code is now, the random number will no longer pick the branches with equal probability.

File: tokio/src/macros/select.rs (hunk 4 above)
Comment (pickfire) [reply tokio/src/macros/select.rs:454]: One issue is modify `select_priv_declare_output_enum`, but it's not really the issue I am concern with. A bigger issue is `let (_, #[cfg(...)] _, ..) = ...;` is not possible.

File: tokio/src/macros/select.rs (hunk 5 above)
Comment (pickfire) [reply tokio/src/macros/select.rs:656]: Why?

Comment (Darksonn) [reply tokio/src/macros/select.rs:656]: To illustrate my point, consider if we have three branches:
```Rust
select! {
    #[cfg(false)]
    _ = branch_0() => {},
    _ = branch_1() => {},
    _ = branch_2() => {},
}
```
Then we have `BRANCHES` equal to 3 and the random number generator will pick either 0, 1 or 2 at equal probability. However, if the generator chooses a 0, then that's turned in to a 1 by the continue. This makes option 1 twice as likely as option 2.

Comment (pickfire) [reply tokio/src/macros/select.rs:656]: Ah true. I didn't thought about that. What should I do for this?

Comment (Darksonn) [reply tokio/src/macros/select.rs:656]: I don't really know what to do about it. I don't think there is an easy fix.

**Comment by Darksonn** [comment]
Closing as the implementation is incorrect and there's no obvious fix. Feel free to reopen or create a new PR if you have a solution.

**Pull request closed without merging**